### PR TITLE
DON-998: Partially implement restriction on creating ad-hoc donation …

### PR DESCRIPTION
**Pull request by bdsl**
…for regular giving campaign or vice-versa

Still need to actuall pull data from SF to make a real distinction, and save it into our DB when SF has that data - see BG2-2711

### CODE REVIEW

File: src/Domain/Campaign.php [hunk 1]
```diff
@@ -77,6 +77,26 @@ public function __toString(): string
         return "Campaign ID #{$this->id}, SFId: {$this->salesforceId}";
     }
 
+    /**
+     * @return bool does this campaign accept ad-hoc giving, i.e. non-regular giving.
+     * @see self::isRegularGiving()
+     */
+    public function isAdHocGiving(): bool
```
Comment (bdsl) [inline src/Domain/Campaign.php:84]: not sure what we want to call the opposite of regular giving. I've gone with ad-hoc giving but there might be a better term.

Comment (bdsl) [reply src/Domain/Campaign.php]: changed to one-off

File: src/Domain/RegularGivingService.php [hunk 2]
```diff
@@ -2,15 +2,160 @@
...
+        );
+        $this->donationService->enrollNewDonation($firstDonation);
+        if (! $firstDonation->isFullyMatched()) {
+            throw new NotFullyMatched(
+                "Donation could not be fully matched, need to match {$firstDonation->getAmount()}," .
+                " only matched {$firstDonation->getFundingWithdrawalTotal()}"
+            );
+        }
+        // @todo-regular-giving - throw if first donation is not fully matched unless donor has said they're OK with
+        //                        that.
```
Comment (NoelLH) [inline src/Domain/RegularGivingService.php:103]:
```suggestion
```

Comment (NoelLH) [reply src/Domain/RegularGivingService.php:103]: And maybe the todo's below? I've slightly lost the thread of which parts were already on `develop`, could be the later parts were done already?

**Review by NoelLH — DISMISSED** [review]
Minor non-blockers inline, LGTM

File: src/Domain/RegularGivingService.php [hunk 3]
```diff
@@ -2,15 +2,158 @@
...
+            mandateSequenceNumber: DonationSequenceNumber::of(1)
+        );
+        $this->donationService->enrollNewDonation($firstDonation);
+        if (! $firstDonation->isFullyMatched()) {
+            throw new NotFullyMatched(
+                "Donation could not be fully matched, need to match {$firstDonation->getAmount()}," .
+                " only matched {$firstDonation->getFundingWithdrawalTotal()}"
+            );
+        }
+        // @todo-regular-giving - collect first donation
```
Comment (bdsl) [inline src/Domain/RegularGivingService.php:102]:
```suggestion
        // @todo-regular-giving - collect first donation (currently created as pending, not collected)
```

**Comment by bdsl** [comment]
Thanks @NoelLH

**Pull request merged**